internal/impl/s2/output.go

```
@@ -0,0 +1,251 @@
+		policy.Count = s2.MaxBatchRecords
+	}
+
+	// This feels sensible to have. Not sure if we should let the user have infinite
+	// retention until a batch of required size is formed.
+	if policy.Period == "" {
+		policy.Period = "5ms"
+	}
+
+	// Validate limits
+
+	if policy.ByteSize > maxBatchBytes {
+		return policy, fmt.Errorf("%w: must not exceed %d", errInvalidBatchingByteSize, s2.MaxBatchBytes)
+	}
+
+	if policy.Count > s2.MaxBatchRecords {
+		return policy, fmt.Errorf("%w: must not exceed %d", errInvalidBatchingCount, s2.MaxBatchRecords)
+	}
+
+	return policy, nil
+}
+
+func newOutputConfig(conf *service.ParsedConfig) (*s2bentobox.OutputConfig, error) {
+	config, err := newConfig(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	stream, err := conf.FieldString(streamField)
+	if err != nil {
+		return nil, err
+	}
+
+	maxInFlight, err := conf.FieldMaxInFlight()
+	if err != nil {
+		return nil, err
+	}
+
+	var fencingToken []byte
+
+	if conf.Contains(fencingTokenField) {
+		field, err := conf.FieldString(fencingTokenField)
+		if err != nil {
+			return nil, err
+		}
+
+		fencingToken, err = base64.StdEncoding.DecodeString(field)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &s2bentobox.OutputConfig{
+		Config:       config,
+		Stream:       stream,
+		MaxInFlight:  maxInFlight,
+		FencingToken: fencingToken,
+	}, nil
+}
+
+type Output struct {
+	inner  *s2bentobox.Output
+	config *s2bentobox.OutputConfig
+	logger *service.Logger
+}
+
+func (o *Output) Connect(ctx context.Context) error {
+	o.logger.Debug("Connecting S2 output")
+
+	inner, err := s2bentobox.ConnectOutput(ctx, o.config)
+	if err != nil {
+		return err
+	}
+
+	// Initialize inner connection.
+	o.inner = inner
+
+	return nil
+}
+
+func (o *Output) WriteBatch(ctx context.Context, batch service.MessageBatch) error {
+	o.logger.Debug("Writing batch to S2")
+
+	recordBatch, err := s2.NewAppendRecordBatch()
+	if err != nil {
+		panic("empty record batch shouldn't error")
+	}
```

Don't think we should be panicking here. Can you rather return an error? If you want the pipeline to terminate I'd recommend returning a `service.ErrEndOfInput`.

Else, if you have more information on what the error is perhaps return a `ErrNotConnected` (to retry connection) or look into a custom error that can be handled by your `ackFn`.

---

This would never error. An empty batch should always be valid. (`s2.NewAppendRecordBatch` takes batches as variadic args, not visible here).

I'll create a variant that doesn't error - `s2.NewEmptyAppendRecordBatch`. Or maybe just do `s2.NewAppendRecordBatch(records...)` and check for error then.

Also, we definitely want `service.ErrEndOfInput` here. This reminds me, there are a bunch of places where this would be helpful.